lib/utils/fakeconn_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"io/ioutil"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+)
+
+// TestCancelableChConn validates that reads from the cancelable channel
+// connection can be canceled by setting a read deadline.
+func TestCancelableChConn(t *testing.T) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+
+	sshConnCh := make(chan sshConn)
+
+	go startSSHServer(t, listener, sshConnCh)
+
+	client, err := ssh.Dial("tcp", listener.Addr().String(), &ssh.ClientConfig{
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	})
+	require.NoError(t, err)
+
+	_, _, err = client.OpenChannel("test", []byte("hello ssh"))
+	require.NoError(t, err)
+
+	select {
+	case sshConn := <-sshConnCh:
+		chConn := NewCancelableChConn(sshConn.conn, sshConn.ch)
+		doneCh := make(chan error)
+		go func() {
+			// Nothing is sent on the channel so this will block until the
+			// read is canceled by the deadline set below.
+			_, err := ioutil.ReadAll(chConn)
+			doneCh <- err
+		}()
+		// Set the read deadline and make sure that the read above is
+		// canceled with a timeout error within a second.
+		chConn.SetReadDeadline(time.Now().Add(200 * time.Millisecond))
```

Comment:
Why not use the same approach with a deadline in the past to avoid waiting?

Reply:
Good idea.